lib/charms/hpc_libs/v0/slurm_ops.py

```diff
@@ -0,0 +1,110 @@
+share and consume charm libraries. They serve to enhance collaboration
+between charmers. Use a charmer's libraries for classes that handle
+integration with their charm.
+
+Bear in mind that new revisions of the different major API versions (v0, v1,
+v2 etc) are maintained independently.  You can continue to update v0 and v1
+after you have pushed v3.
+
+Markdown is supported, following the CommonMark specification.
+"""
+
+import logging
+import functools
+import subprocess
+import yaml
+import base64
+import enum
+
+
+_logger = logging.getLogger(__name__)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "541fd767f90b40539cf7cd6e7db8fabf"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+def _snap(*args) -> str:
+    """Control snap by via executed `snap ...` commands.
+
+    Raises:
+        subprocess.CalledProcessError: Raised if snap command fails.
+    """
+    cmd = ["snap", *args]
+    _logger.debug("snap: Executing command %s", cmd)
+    try:
+        return subprocess.check_output(cmd, text=True)
+    except subprocess.CalledProcessError as e:
+        _logger.error("snap command failed: %s", e)
+        raise
+
+_get_config = functools.partial(_snap, "get", "slurm")
+_set_config = functools.partial(_snap, "set", "slurm")
+
+class Service(enum.Enum):
+    SLURMD = "slurmd"
+    SLURMCTLD = "slurmctld"
+    SLURMDBD = "slurmdbd"
+    SLURMRESTD = "slurmrestd"
+
+    @property
+    def config_name(self) -> str:
+        if self is Service.SLURMCTLD:
+            return "slurm"
+        return self.value
+
+class SlurmManager:
+    def __init__(self, service: Service):
+        self._service = service
+
+    def install(self):
+        # TODO: Pin slurm to the stable channel
+        _snap("install", "slurm", "--channel", "edge")
+
+    def start(self):
+        _snap("start", "--enable", f"slurm.{self._service.value}")
+
+    def restart(self):
+        _snap("restart", f"slurm.{self._service.value}")
+
+    def disable(self):
+        _snap("stop", "--disable", f"slurm.{self._service.value}")
+
+    def set_config(self, key: str, value: str):
+        _set_config(f"{self._service.config_name}.{key}={value}")
+
+    def get_config(self, key: str) -> str:
+        _get_config(f"{self._service.config_name}.{key}")
+
+    def set_munge_key(self, key: bytes):
+        # TODO: use `slurm.setmungekey` when implemented
+        # subprocess.run(["slurm.setmungekey"], stdin=key)
+        key = base64.encodebytes(key)
+        _set_config(f"munge.key={key}")
+
+    def get_munge_key(self) -> bytes:
+        # TODO: use `slurm.setmungekey` when implemented
+        # key = subprocess.run(["slurm.getmungekey"])
+        key = _get_config("munge.key")
+        return base64.decodebytes(key.encode())
```

Comment:
Do we also want to have a method to generate a munge key? What about the jwt key?

Reply:
Good point - this is something that we should do.

The Slurm snap already generates a key at installation, and you would use `mungekey -c` to create a new key using the snap. We should have the snap generate the jwt, and create a new one when necessary.

What do you think about that?

Reply:
Yep, mirroring the munge key behaviour sounds good to me.